jbmc/unit/java_bytecode/goto_program_generics/mutually_recursive_generics.cpp

```diff
@@ -0,0 +1,310 @@
+            "java::java.lang.Byte",
+            "java::java.lang.Character");
+
+          THEN("`rotate` has itself a field `rotate` ... ")
+          {
+            const auto &rotate_rec_field =
+              require_goto_statements::require_struct_component_assignment(
+                rotate_field,
+                {},
+                "rotate",
+                "java::Three",
+                {},
+                entry_point_code);
+            has_x_u_e_fields(
+              rotate_rec_field,
+              "java::java.lang.Character",
+              "java::java.lang.Integer",
+              "java::java.lang.Byte");
+          }
+          THEN("`rotate` has also a field `normal` ... ")
+          {
+            const auto &rotate_normal_field =
+              require_goto_statements::require_struct_component_assignment(
+                rotate_field,
+                {},
+                "normal",
+                "java::Three",
+                {},
+                entry_point_code);
+            has_x_u_e_fields(
+              rotate_normal_field,
+              "java::java.lang.Integer",
+              "java::java.lang.Byte",
+              "java::java.lang.Character");
+          }
+        }
+        THEN("`example2` has field `normal`")
+        {
+          const auto &normal_field =
+            require_goto_statements::require_struct_component_assignment(
+              example2_field,
+              {},
+              "normal",
+              "java::Three",
+              {},
+              entry_point_code);
+          has_x_u_e_fields(
+            normal_field,
+            "java::java.lang.Byte",
+            "java::java.lang.Character",
+            "java::java.lang.Integer");
+          THEN("`normal` has itself a field `normal`")
+          {
+            const auto &normal_rec_field =
+              require_goto_statements::require_struct_component_assignment(
+                example2_field,
+                {},
+                "normal",
+                "java::Three",
+                {},
+                entry_point_code);
+            has_x_u_e_fields(
+              normal_rec_field,
+              "java::java.lang.Byte",
+              "java::java.lang.Character",
+              "java::java.lang.Integer");
+          }
+          THEN("`normal` has also a field `rotate`")
+          {
+            const auto &normal_rotate_field =
+              require_goto_statements::require_struct_component_assignment(
+                example2_field,
+                {},
+                "rotate",
+                "java::Three",
+                {},
+                entry_point_code);
+            has_x_u_e_fields(
+              normal_rotate_field,
+              "java::java.lang.Integer",
+              "java::java.lang.Byte",
+              "java::java.lang.Character");
+          }
+        }
+      }
+    }
+    THEN("The Object has a field `example3` of type `Outer<Boolean, Byte>`")
```

Comment:
This should be `Outer<Boolean, Byte, Short>`, and fields for `inner` are also not quite right - the type of `o` has also three parameters and test for `u` is missing.

Reply:
I'm not quite sure how this test is passing actually :)

Reply:
In `THEN`, the type should be `Outer<Boolean, Byte, Short>`, that's the full type of `example3`. This is the only thing that is wrong, so now I see how this is passing, the test for `inner` is indeed correct. What got me confused is that the correct instantiation is tested via checking the types of fields, not the expression type (`java_generic_typet` with the expected list of generic variables). I guess that would be a completely different test, in a later stage than goto I guess? In any case, the only thing that doesn't match is the comment in `THEN`.